nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDBRecord.java

```diff
@@ -0,0 +1,283 @@
+
+            HashMap<String, Tablet> tablets = generateTablets(schema, prefix, maxRowNumber);
+            DateTimeFormatter format = null;
+
+            Record record;
+
+            while ((record = recordReader.nextRecord()) != null) {
+                long timestamp = getTimestamp(schema, record);
+                boolean isFulled = false;
+
+                for (Map.Entry<String, Tablet> entry : tablets.entrySet()) {
+                    String device = entry.getKey();
+                    Tablet tablet = entry.getValue();
+                    int rowIndex = tablet.rowSize++;
+
+                    tablet.addTimestamp(rowIndex, timestamp);
+                    List<MeasurementSchema> measurements = tablet.getSchemas();
+                    for (MeasurementSchema measurement : measurements) {
+                        String id = measurement.getMeasurementId();
+                        TSDataType type = measurement.getType();
+                        Object value = getValue(record.getValue(id), type);
+                        tablet.addValue(id, rowIndex, value);
+                    }
+                    isFulled = tablet.rowSize == tablet.getMaxRowNumber();
+                }
+                if (isFulled) {
+                    if (aligned) {
+                        session.get().insertAlignedTablets(tablets);
+                    } else {
+                        session.get().insertTablets(tablets);
+                    }
+                    tablets.values().forEach(tablet -> tablet.reset());
+                }
+            }
+
+            AtomicBoolean hasRest = new AtomicBoolean(false);
+            tablets.forEach(
+                    (device, tablet) -> {
+                        if (hasRest.get() == false && tablet.rowSize != 0) {
+                            hasRest.set(true);
+                        }
+                    });
+            if (hasRest.get()) {
+                if (aligned) {
+                    session.get().insertAlignedTablets(tablets);
+                } else {
+                    session.get().insertTablets(tablets);
+                }
+            }
+        } catch (Exception e) {
+            getLogger().error("Processing failed {}", flowFile, e);
+            processSession.transfer(flowFile, REL_FAILURE);
+            e.printStackTrace();
+            return;
+        }
+        processSession.transfer(flowFile, REL_SUCCESS);
+    }
+
+    private IoTDBSchema getSchema(String property, RecordReader recordReader) throws Exception {
+        ValidationResult result;
+        IoTDBSchema schema;
+        result =
+                property != null
+                        ? validateSchemaAttribute(property)
+                        : validateSchema(recordReader.getSchema());
+
+        if (!result.getKey()) {
+            getLogger().error("The property `schema` has an error: {}", result.getValue());
+            throw new Exception();
```

Comment:
An empty `Exception` should not be thrown. This should be a more specific type of Exception, such as IllegalStateException, and a message should always be included.

Reply:
@lizhizhou Please adjust the exception type and include a message. Using `ProcessException` is a good option.